File: currency_favorites.py
# currency_favorites.py
import sys
import json
import os
import requests
from datetime import datetime

FAVORITES_FILE = "favorites.json"
API_URL = "https://api.frankfurter.app/latest?from="

class CurrencyMonitor:
    def __init__(self):
        self.favorites = []
        self.load_favorites()

    def load_favorites(self):
        if os.path.exists(FAVORITES_FILE):
            with open(FAVORITES_FILE, "r") as f:
                self.favorites = json.load(f)
        else:
            self.favorites = [["USD", "EUR"], ["USD", "JPY"]]  # default

    def save_favorites(self):
        with open(FAVORITES_FILE, "w") as f:
            json.dump(self.favorites, f, indent=2)

    def add_pair(self, base, target):
        pair = [base.upper(), target.upper()]
        if pair not in self.favorites:
            self.favorites.append(pair)
            self.save_favorites()
            print(f"✅ Added {base}/{target}")
        else:
            print(f"Pair {base}/{target} already in favorites.")

    def remove_pair(self, base, target):
        pair = [base.upper(), target.upper()]
        if pair in self.favorites:
            self.favorites.remove(pair)
            self.save_favorites()
            print(f"✅ Removed {base}/{target}")
        else:
            print(f"Pair {base}/{target} not found.")

    def get_rates(self):
        # Group by base currency to reduce API calls
        rates = {}
        for base, target in self.favorites:
            if base not in rates:
                # Fetch all rates from this base
                url = API_URL + base
                try:
                    resp = requests.get(url)
                    resp.raise_for_status()
                    data = resp.json()
                    rates[base] = data["rates"]
                except Exception as e:
                    print(f"Error fetching rates for {base}: {e}", file=sys.stderr)
                    rates[base] = {}
        return rates

    def list_rates(self):
        if not self.favorites:
            print("No favorite pairs.")
            return
        rates = self.get_rates()
        print("\n💱 Favorite Currency Pairs\n")
        for base, target in self.favorites:
            if target in rates.get(base, {}):
                rate = rates[base][target]
                print(f"{base}/{target}: {rate:.4f}")
            else:
                print(f"{base}/{target}: N/A")
        print(f"\nUpdated: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")

def main():
    app = CurrencyMonitor()
    if len(sys.argv) < 2:
        app.list_rates()
    else:
        cmd = sys.argv[1].lower()
        if cmd == "list":
            app.list_rates()
        elif cmd == "add" and len(sys.argv) == 4:
            app.add_pair(sys.argv[2], sys.argv[3])
        elif cmd == "remove" and len(sys.argv) == 4:
            app.remove_pair(sys.argv[2], sys.argv[3])
        elif cmd == "watch":
            import time
            interval = 60
            if len(sys.argv) >= 4 and sys.argv[2] == "--interval":
                interval = int(sys.argv[3])
            print(f"Watching every {interval}s. Press Ctrl+C to stop.")
            try:
                while True:
                    app.list_rates()
                    time.sleep(interval)
            except KeyboardInterrupt:
                print("\nStopped.")
        else:
            print("Usage: currency_favorites.py [list|add BASE TARGET|remove BASE TARGET|watch [--interval N]]")

if __name__ == "__main__":
    main()

File: currency_favorites.js
// currency_favorites.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');

const FAVORITES_FILE = 'favorites.json';
const API_URL = 'https://api.frankfurter.app/latest?from=';

function loadFavorites() {
    if (fs.existsSync(FAVORITES_FILE)) {
        try {
            const data = JSON.parse(fs.readFileSync(FAVORITES_FILE));
            return data.pairs || [['USD', 'EUR'], ['USD', 'JPY']];
        } catch (e) {
            return [['USD', 'EUR'], ['USD', 'JPY']];
        }
    }
    return [['USD', 'EUR'], ['USD', 'JPY']];
}

function saveFavorites(pairs) {
    fs.writeFileSync(FAVORITES_FILE, JSON.stringify({ pairs }, null, 2));
}

async function getRates(base) {
    const url = API_URL + base;
    const resp = await fetch(url);
    if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
    const data = await resp.json();
    return data.rates;
}

function listRates(pairs) {
    if (pairs.length === 0) {
        console.log('No favorite pairs.');
        return;
    }
    // Group by base
    const grouped = {};
    for (const [base, target] of pairs) {
        if (!grouped[base]) grouped[base] = [];
        grouped[base].push(target);
    }
    console.log('\n💱 Favorite Currency Pairs\n');
    Promise.all(Object.entries(grouped).map(async ([base, targets]) => {
        try {
            const rates = await getRates(base);
            for (const target of targets) {
                if (rates[target] !== undefined) {
                    console.log(`${base}/${target}: ${rates[target].toFixed(4)}`);
                } else {
                    console.log(`${base}/${target}: N/A`);
                }
            }
        } catch (e) {
            console.error(`Error fetching rates for ${base}: ${e.message}`);
        }
    })).then(() => {
        console.log(`\nUpdated: ${new Date().toISOString().slice(0,19).replace('T',' ')}`);
    }).catch(console.error);
}

function main() {
    const args = process.argv.slice(2);
    if (args.length === 0) {
        const pairs = loadFavorites();
        listRates(pairs);
        return;
    }
    const cmd = args[0].toLowerCase();
    if (cmd === 'list') {
        const pairs = loadFavorites();
        listRates(pairs);
    } else if (cmd === 'add' && args.length === 3) {
        const base = args[1].toUpperCase();
        const target = args[2].toUpperCase();
        const pairs = loadFavorites();
        const newPair = [base, target];
        if (!pairs.some(p => p[0] === base && p[1] === target)) {
            pairs.push(newPair);
            saveFavorites(pairs);
            console.log(`✅ Added ${base}/${target}`);
        } else {
            console.log(`Pair ${base}/${target} already in favorites.`);
        }
    } else if (cmd === 'remove' && args.length === 3) {
        const base = args[1].toUpperCase();
        const target = args[2].toUpperCase();
        let pairs = loadFavorites();
        const filtered = pairs.filter(p => !(p[0] === base && p[1] === target));
        if (filtered.length < pairs.length) {
            saveFavorites(filtered);
            console.log(`✅ Removed ${base}/${target}`);
        } else {
            console.log(`Pair ${base}/${target} not found.`);
        }
    } else if (cmd === 'watch') {
        let interval = 60;
        const idx = args.indexOf('--interval');
        if (idx !== -1 && args.length > idx + 1) {
            interval = parseInt(args[idx + 1]);
        }
        console.log(`Watching every ${interval}s. Press Ctrl+C to stop.`);
        const pairs = loadFavorites();
        setInterval(() => listRates(pairs), interval * 1000);
        listRates(pairs);
    } else {
        console.log('Usage: currency_favorites [list|add BASE TARGET|remove BASE TARGET|watch [--interval N]]');
    }
}

main().catch(console.error);
